networkx/algorithms/bipartite/extendability.py

```diff
@@ -0,0 +1,115 @@
+""" Provides a function for computing the extendability of a graph which is
+undirected, simple, connected and bipartite and contains at least one perfect matching."""
+
+
+import networkx as nx
+from networkx.utils import not_implemented_for
+
+__all__ = ["maximal_extendability"]
+
+
+@not_implemented_for("directed")
+@not_implemented_for("multigraph")
+def maximal_extendability(G):
+    """Computes the extendability of a graph.
+
+    The extendability of a graph is defined as the maximum $k$ for which `G`
+    is $k$-extendable. Graph `G` is $k$-extendable if and only if `G` has a
+    perfect matching and every set of $k$ independent edges can be extended
+    to a perfect matching in `G`.
+
+    Parameters
+    ----------
+
+    G : NetworkX Graph
+        A fully-connected bipartite graph without self-loops
+
+    Returns
+    -------
+
+    extendability : int
+
+    Raises
+    ------
+
+    NetworkXError
+       If the graph G contains self-loops.
+       If the graph G is disconnected.
+       If the graph G is not bipartite.
+       If the graph G does not contain a perfect matching.
+       If the residual graph of G is not strongly connected.
+
+    Notes
+    -----
+
+    Definition:
+    Let G be a simple, connected, undirected and bipartite graph with a perfect
+    matching M and bipartition (U,V). The residual graph of G, denoted by $G_M$,
+    is the graph obtained from G by directing the edges of M from V to U and the
+    edges that do not belong to M from U to V.
+
+    Lemma([1]_):
+    Let M be a perfect matching of G. G is $k$-extendable if and only if its residual
+    graph $G_M$ is strongly connected and there are $k$ vertex-disjoint directed
+    paths between every vertex of U and every vertex of V.
+
+    Assuming that input graph G is undirected, simple, connected, bipartite and contains
+    a perfect matching M, this function constructs the residual graph $G_M$ of G and
+    returns the minimum value among the maximum vertex-disjoint directed paths between
+    every vertex of U and every vertex of V in $G_M$. By combining the definitions
+    and the lemma, this value represents the extendability of the graph G.
+
+    Time complexity O($n^3$ $m^2$)) where $n$ is the number of vertices
+    and $m$ is the number of edges.
+
+    References
+    ----------
+
+    ..[1] "A polynomial algorithm for the extendability problem in bipartite graphs",
+          J. Lakhal, L. Litzler, Information Processing Letters, 1998.
+    ..[2] "On n-extendible graphs", M. D. Plummer, Discrete Mathematics, 31:201–210, 1980
+          https://doi.org/10.1016/0012-365X(80)90037-0
+
+    """
+
+    if any(nx.nodes_with_selfloops(G)):
+        raise nx.NetworkXError("Graph G has selfloops")
```

Comment:
In the process of adding a test for this I noticed a corner case; namely when the *only* node containing self edges is False-y:

```python
>>> G = nx.ladder_graph(3)
>>> G.add_edge(0, 0)
>>> list(nx.nodes_with_selfloops(G))
[0]
# Therefore...
>>> any(nx.nodes_with_selfloops(G))
False
```

Is this explicit check even needed? Aren't graphs that have self-edges non-bipartite by definition?

Reply:
Ack!  Good Catch!  And even better resolution!  
self-loops make it not bipartitie!  So, we don't need this check.

Thanks:)  I've removed that faulty check -- and not replaced it since bipartite checks that.